### PR TITLE
feat(cloudrun): Adding cloudrun provider functionality in orca.

### PR DESCRIPTION
**Pull request by sanopsmx**
This is part of https://github.com/spinnaker/governance/issues/302

Adding cloudrun provider functionality in orca.

### CODE REVIEW

**Comment by mattgogerly** [comment]
@sanopsmx any chance `CloudrunServerGroupCreator` could be rewritten in Java? LGTM otherwise

**Comment by sanopsmx** [comment]
> @sanopsmx any chance `CloudrunServerGroupCreator` could be rewritten in Java? LGTM otherwise

Modified the code from groovy to java.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/strategy/DetermineSourceServerGroupTask.groovy [hunk 1]
```diff
@@ -49,6 +51,15 @@ class DetermineSourceServerGroupTask implements RetryableTask {
...
   TaskResult execute(StageExecution stage) {
+    ObjectMapper yamlReader = new ObjectMapper(new YAMLFactory());
+    if (stage.getContext() != null && stage.getContext().get("configFiles") != null
+        && (!((List) stage.getContext().get("configFiles")).isEmpty())) {
+      Object yamlObj = yamlReader.readValue(stage.getContext().get("configFiles").get(0), Object.class);
+      if (yamlObj.get("metadata") != null && yamlObj.get("metadata").get("labels") != null
+          && yamlObj.get("metadata").get("labels").get("cloud.googleapis.com/location") != null) {
+        stage.getContext().put("region", yamlObj.get("metadata").get("labels").get("cloud.googleapis.com/location"))
+      }
+    }
```
Comment (mattgogerly) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/strategy/DetermineSourceServerGroupTask.groovy:62]: I'm not convinced this is the right place for this. `DetermineSourceServerGroups` is a generic task used by lots of different providers, so having custom stuff for one provider in it seems weird.

Can we not extend this and override the `execute` method for Cloudrun?

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForRequiredInstancesDownTask.groovy [hunk 2]
```diff
@@ -42,7 +42,10 @@ public class WaitForRequiredInstancesDownTask extends AbstractInstancesCheckTask
     if (interestingHealthProviderNames != null && interestingHealthProviderNames.isEmpty()) {
       return true
     }
-
+    String cloudProvider = getCloudProvider(stage);
+    if ("cloudrun".equals(cloudProvider)) {
+      return true;
+    }
```
Comment (mattgogerly) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/WaitForRequiredInstancesDownTask.groovy:48]: Same comment as below - can we just extend this for Cloudrun?

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/strategy/DetermineSourceServerGroupTask.groovy [hunk 3]
```diff
@@ -49,6 +50,7 @@ class DetermineSourceServerGroupTask implements RetryableTask {
 
   @Override
   TaskResult execute(StageExecution stage) {
+
```
Comment (mattgogerly) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/kato/pipeline/strategy/DetermineSourceServerGroupTask.groovy:53]: nit: space not needed

File: orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/providers/cloudrun/CloudrunServerGroupCreator.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CloudrunServerGroupCreator implements ServerGroupCreator {
+
+  @Autowired ObjectMapper objectMapper;
```
Comment (mattgogerly) [inline orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/providers/cloudrun/CloudrunServerGroupCreator.java:39]: Constructor injection is preferred over property injection

File: orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/kato/pipeline/strategy/CloudrunSourceServerGroupTask.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    if (stage.getContext() != null
+        && stage.getContext().get("configFiles") != null
+        && (!((List) stage.getContext().get("configFiles")).isEmpty())) {
+      Map<String, Object> contextMap = stage.getContext();
+      String configStr = (String) ((List) contextMap.get("configFiles")).get(0);
+      LinkedHashMap<String, Object> yamlMap = null;
+      try {
+        yamlMap = yamlReader.readValue(configStr, LinkedHashMap.class);
+        if (yamlMap.get("metadata") != null
+            && ((LinkedHashMap<String, Object>) yamlMap.get("metadata")).get("labels") != null
```
Comment (mattgogerly) [inline orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/kato/pipeline/strategy/CloudrunSourceServerGroupTask.java:48]: Is the cast to LinkedHashMap necessary here or can it just be Map?

**Review by mattgogerly — COMMENTED** [review]
Couple of comments. Can you add some tests for the new functionality as well please?

File: orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/providers/cloudrun/CloudrunServerGroupCreator.java (hunk 4 above)
Comment (sanopsmx) [reply orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/clouddriver/tasks/providers/cloudrun/CloudrunServerGroupCreator.java]: done

File: orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/kato/pipeline/strategy/CloudrunSourceServerGroupTask.java (hunk 5 above)
Comment (sanopsmx) [reply orca-clouddriver/src/main/java/com/netflix/spinnaker/orca/kato/pipeline/strategy/CloudrunSourceServerGroupTask.java]: done

**Comment by sanopsmx** [comment]
> Couple of comments. Can you add some tests for the new functionality as well please?

Added test cases and sent it for review.

**Review by skandragon — APPROVED** [review]
LGTM

**Pull request merged**